File: Data Structures & Algorithms/permutations/submission-1.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        self.res = []
        self.backTrack([], nums, [False]*len(nums))
        return self.res
    
    def backTrack(self, perm, nums, visit):
        if len(perm) == len(nums):
            self.res.append(perm[:])
            return
        for i in range(len(nums)):
            if not visit[i]:
                perm.append(nums[i])
                visit[i] = True
                self.backTrack(perm, nums, visit)
                perm.pop()
                visit[i] = False

File: Data Structures & Algorithms/permutations-ii/submission-1.go
func permuteUnique(nums []int) [][]int {
    res := make(map[string][]int)
    perm := []int{}

    var dfs func()
    dfs = func(){
        if len(perm) == len(nums){
            temp := make([]int, len(perm))
            copy(temp, perm)
            key := fmt.Sprint(temp)
            res[key] = temp
            return
        }

        for i:=0; i<len(nums); i++{
            if nums[i] != -1001{
                temp := nums[i]
                perm = append(perm, nums[i])
                nums[i] = -1001
                dfs()
                nums[i] = temp
                perm = perm[:len(perm)-1]
            }
        }
    }
    dfs()
    result := [][]int{}
    for _,v := range res{
        result = append(result,v)
    }
    return result
}
